nutils/evaluable.py

```diff
@@ -757,10 +757,30 @@ def _assparse(self):
       return chunks
     return _assparse
 
+  class _evalf_checker:
+    def __init__(self, orig):
+      self.evalf_obj = getattr(orig, '__get__', lambda *args: orig)
+    def __get__(self, instance, owner):
+      evalf = self.evalf_obj(instance, owner)
+      @functools.wraps(evalf)
+      def evalf_with_check(*args, **kwargs):
+        res = evalf(*args, **kwargs)
+        if hasattr(instance, 'dtype'):
+          assert asdtype(res.dtype) == instance.dtype, ((instance.dtype, res.dtype), instance, res)
+        try:
+          t = Tuple(instance.shape).eval()
+        except: pass
+        else:
+          assert t == res.shape, ((instance.shape, res.shape), instance, res)
```

Comment:
Instead of the `try ... except ... else ...` I suggest something like (unverified)

```python
assert res.ndim == self.ndim
assert all(n == m for n, m in zip(res.shape, self.shape) if isinstance(m, int))
```

This allows partial checking of the shape if only some of the axes have a variable length. This is not uncommon: the what-used-to-be-the-points-axis is often the only axis that has a variable length.